### DISCUSSION TITLE
Noise / uncertainty injection

### DISCUSSION THREAD

**Discussion by beckynevin** [deepskies/DeepUQ#15] [General]
The eventual goal is to have a utility that will wrap around something like deeputils (utils on utils) to inject uncertainty and then quantify how this is accounted for by various ML and statistical tools on the backend. Here let's discuss the error injection part of this. 

- How has this been done in other work? 
- How do we want to do this here?
- Do we want to explore the pendulum parameters separately and how error propagates? Do we want to look at any correlated error?
- How do we choose the levels of error propagation? I recommend we do a similar thing to deeplyuncertain, where we go with 1%, 10%, maybe 50% error levels and explore how they affect the predicted error.

This discussion is to discuss the above.

Additionally, the error propagation notebook example runs through how error is propagated by the deepbench utility and how this compares to our analytical expectation of error propagation. At present, error injected on theta is matching with my expectation of what the analytical error should be but not on L or a_g.

**Comment by beckynevin** [comment]
@bnord do you have any references for error injection? 

So far the best source I've found is this review on UQ by Abdar+2021 [https://www.sciencedirect.com/science/article/pii/S1566253521001081](https://www.sciencedirect.com/science/article/pii/S1566253521001081).

There's also this thesis which talks a bit about noise injection: [https://core.ac.uk/download/pdf/237398383.pdf](https://core.ac.uk/download/pdf/237398383.pdf) 

I'd like to find at least one source like Caldeira & Nord that does some sort of analytic propagation to compare their output predicted uncertainty to the analytic expectation. I'm wondering if maybe I'm using the wrong words to search for this. Maybe this field is 'calibration?'

**Reply by bnord** [reply]
Joao and I couldn't really find any sources that did what we did, which I would characterize as:
1. physics equations and models
2. propagate error 
3. compare analytic stuff to ML stuff


Are you looking for sources that did those 3 things or a subset of those 3 things?
Do I understand your question correctly?

**Reply by beckynevin** [reply]
I would say I'm mostly looking for #2 and #3 hopefully together but also I'm looking for them separately.

**Reply by bnord** [reply]
there might be some combo's out there for #2 and #3. One issue is that ML error stuff is often done on natural images, so noise doesn't typically get characterized or propagated in a mathematically robust way.

**Reply by bnord** [reply]
what's your list of search of terms? Maybe I can add to that?

**Reply by beckynevin** [reply]
Alright, let's see. This is my best lead so far - an excellent paragraph summarizing how people are talking about error injection from [Nagl 2022](https://www.frontiersin.org/articles/10.3389/fams.2022.1076083/full#:~:text=A%20first%20intuitive%20way%20to,network%20architectures%20containing%20many%20parameters.)

> An important aspect, to which the machine learning LGD literature has not yet paid attention, is the associated uncertainty around estimates and predictions[2](https://www.frontiersin.org/articles/10.3389/fams.2022.1076083/full#note2). Commonly, we can define two types of uncertainty, aleatoric and epistemic [[33](https://www.frontiersin.org/articles/10.3389/fams.2022.1076083/full#B33)]. Following Gawlikowski et al. [[34](https://www.frontiersin.org/articles/10.3389/fams.2022.1076083/full#B34)], aleatoric uncertainty is the uncertainty in the data itself that can not be reduced and is therefore also known as irreducible or data uncertainty. In classical statistics, this type of uncertainty is for example represented by ϵ in the linear regression framework. Epistemic uncertainty refers to the uncertainty of a model due to the (limited) sample size. This uncertainty can be reduced by increasing the sample size on which the model is trained and is therefore also known as reducible or model uncertainty [[34](https://www.frontiersin.org/articles/10.3389/fams.2022.1076083/full#B34)]. In a linear regression setting, epistemic uncertainty is, accounted for by the standard errors of the beta coefficients. Given a larger sample size, the standard errors should decrease. Recently, the literature on uncertainty estimation has grown rapidly as outlined in a survey article by Gawlikowski et al. [[34](https://www.frontiersin.org/articles/10.3389/fams.2022.1076083/full#B34)].

**Reply by beckynevin** [reply]
I'm reading over this source again. It's interesting that to them, 'aleatoric' is simply a perturbation on the linear regression. I think this means `y = mx + b + epsilon`, whereas 'epistemic' is the standard error of the beta coefficients, which I think means the combination of coefficients m and b. I'm guessing this means in their case, the network is taking in y and x as the 'data' and outputting m and b? If that's the case, then I think the error injection in epsilon is a measurement error in a way that is applied to either y or x (you could think about it either way since both x and y are the input to the network). This is also helping me reframe how I think about forward versus backward inference.

**Comment by beckynevin** [comment]
I also found this source that is leading me in some useful directions. I think our relevant terminology for searching is stuff like "calibration" [Gawlikowski 2023](https://link.springer.com/article/10.1007/s10462-023-10562-9)

> However, estimating the predictive uncertainty is not sufficient for safe decision-making. It is also crucial to assure that the uncertainty estimates are reliable. To this end, the calibration property (the degree of reliability) of DNNs has been investigated and re-calibration methods have been proposed (Guo et al. [2017](https://link.springer.com/article/10.1007/s10462-023-10562-9#ref-CR107); Wenger et al. [2020](https://link.springer.com/article/10.1007/s10462-023-10562-9#ref-CR349); Zhang et al. [2020](https://link.springer.com/article/10.1007/s10462-023-10562-9#ref-CR371)) to obtain reliable (well-calibrated) uncertainty estimates.

**Comment by beckynevin** [comment]
I found a great thesis I'm going to read through about noise injection methods - [https://pastel.hal.science/tel-03255379/document](https://pastel.hal.science/tel-03255379/document)
<img width="619" alt="image" src="https://github.com/deepskies/DeepUQ/assets/26393168/8fe74530-2492-4f98-be10-2ae50a15982c">

**Reply by beckynevin** [reply]
Unfortunately, I think they mean NN weights when they say 'model parameters', not what we're used to thinking of as model parameters.

**Reply by beckynevin** [reply]
After our discussion today @bnord I'm now thinking about 'adding noise to the model parameters' in a different light. I do think they mean injecting directly into the NN weights (latent parameters, z). I'm interested in thinking about this more when I come back and digging more into the differences in how 'parameters' are considered in deterministic models versus inference models.
